core/src/main/java/bio/terra/pearl/core/service/publishing/PublishingSupport.java

```diff
@@ -0,0 +1,65 @@
+package bio.terra.pearl.core.service.publishing;
+
+import bio.terra.pearl.core.model.BaseEntity;
+import bio.terra.pearl.core.model.Versioned;
+import bio.terra.pearl.core.model.portal.PortalEnvironment;
+import bio.terra.pearl.core.model.publishing.*;
+import bio.terra.pearl.core.model.study.StudyEnvironment;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+public interface PublishingSupport {
+    public static final List<String> CONFIG_IGNORE_PROPS = List.of("id", "createdAt", "lastUpdatedAt", "class",
+            "studyEnvironmentId", "portalEnvironmentId", "emailTemplateId", "emailTemplate",
+            "consentFormId", "consentForm", "surveyId", "survey", "versionedEntity", "trigger");
+    public void loadForDiffing(PortalEnvironment portalEnv);
+    public void loadForDiffing(StudyEnvironment studyEnv);
+    public void updateDiff(PortalEnvironment sourceEnv, PortalEnvironment destEnv, PortalEnvironmentChange change);
+    public void updateDiff(StudyEnvironment sourceEnv, StudyEnvironment destEnv, StudyEnvironmentChange change);
+    public void applyDiff(PortalEnvironment sourceEnv, PortalEnvironment destEnv, PortalEnvironmentChange change);
+    public void applyDiff( StudyEnvironmentChange change, StudyEnvironment destEnv, PortalEnvironment destPortalEnv);
```

Comment:
Just for my own understanding: why are there separate portalenvironment and studyenvironment methods? It feels like some things are publishable only at a portal level and some only at a study level, but it seems like that isn't the case for triggers, which update for both?

Reply:
that's exactly right.  This was the only way I could think of structuring it so that the entity services didn't have to have the logic for iterating across studies within a portal.